### DISCUSSION TITLE
Ensuring Nullable Properties Aren't Null

### DISCUSSION THREAD

**Discussion by vvanslyke** [prisma/prisma#12358] [Q&A]
Hi Prisma folks!  I am utilizing Typescript Interfaces to ensure I'm retrieving and mapping data correctly before I send it off to another service to be manipulated more and hitting a snag with ensuring nullable properties aren't null.

I am doing an update of many `Unit` objects to add `vendorId` to them.  
I verify that the `vendorId` is associated with a valid `Vendor.`  
`vendorId` is nullable on the `Unit` object, so when I am doing a `findMany` to retrieve the updated `Unit` objects with an associated `Vendor` property, I am still getting an `| null` tacked on to the end of the `Vendor` relationship.  

When I try to fit my `updated_units` into an `IUpdatedUnits` interface, I am getting an error because the `| null` on the `Vendor` relationship is still there.  

I attempted specifying the exact `vendorId` within my where clause in the `findMany`
`where: { id: { in: unitIds }, vendorId: vendor.id },`
and I also tried to utilize `NOT`
`where: { id: { in: unitIds }, NOT: [{ vendor: null }]}, `
but that did not remove the `| null`.

I have the `!updated_units` and `some` check before my service call as well, but it doesn't fix the issue and seems redundant.

Is there something I'm missing with my query in order to tell Prisma that even though we're working with a nullable property, that I have accounted for it being there and to remove that type checking?

🔴  Error is here

Thanks!

Vinny

// ----------Code------------ 

Unit Schema
```prisma
model Unit {
  id                       String                    @id @default(uuid())
  lp                       String                    @unique
  vendor                   Vendor?                   @relation(fields: [vendorId], references: [id])
  vendorId                 String?
}
```

Interfaces
```typescript
export interface IUpdatedUnits {
  lp: string;
  vendor: { vendorIdentifier: string };
}

export interface IGoalObject {
  lps: String[];
  vendor_identifier: string;
}
```

Code
```typescript
{
  const unitIds: String[] = updateParams.unitIds;
  const vendorId: string = updateParams.vendorId;
  const vendor = await this.prismavendor.findUnique({ where: { id: vendorId } });
  
  if (!vendor) {
    return false;
  }
  
  await this.prisma.unit.updateMany({
    where: { id: { in: unitIds } },
    data: { vendor.id }
  });
  
  const updated_units = await this.prisma.unit.findMany({
    where: { id: { in: unitIds }, vendorId: vendor.id }
    select: {
      lp: true,
      vendor: { select: { vendorIdentifier: true } }
    }
  });
  
  if (!updated_units || (updated_units.some((unit) => {return unit.vendor === null;}))) return false;
  
  await this.service.method(this.mappedUnits(updated_units🔴  ));
}

mappedUnits(units: IUpdatedUnits[]): IGoalObject {
    const allLps: string[] = units.map((unit) => {
      return unit.lp;
    });
    return {
      lps: allLps,
      vendor_identifier: units[0].vendor.vendorIdentifier
    };
  }
```

Error
```
const updated_units: {
    lp: string;
    vendor: {
        vendorIdentifier: string;
    } | null;
}[]

"Argument of type '{ lp: string; vendor: { vendorIdentifier: string; } | null; }[]' is not assignable to parameter of type 'IUpdatedUnits[]'.
  Type '{ lp: string; vendor: { vendorIdentifier: string; } | null; }' is not assignable to type 'IUpdatedUnits'.
    Types of property 'vendor' are incompatible.
      Type '{ vendorIdentifier: string; } | null' is not assignable to type '{ vendorIdentifier: string; }'.
        Type 'null' is not assignable to type '{ vendorIdentifier: string; }'.ts(2345)"
```

**Comment by pantharshit00** [comment]
Hello @vvanslyke, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

If this is still relevant, can you please also share the `Vendor` model here? That way I will be able to reproduce this locally without making any assumptions.

**Reply by vvanslyke** [reply]
Not directly relevant still, but I'm still curious about the answer.  This is an applicable representation of the `Vendor` model for the scenario.    

```prisma
model Vendor {
  id                                     String                    @id @default(uuid())
  vendorIdentifier                       String                    
}
```
The biggest point was that I thought I had shown prisma that I had done null checks on my own and it did not catch that the object was not null.  

Thanks for reaching out!

**Reply by pantharshit00** [reply]
I would so something like this to ensure types are casted properly using a type predicate:
```ts
 const unitIds: string[] = updateParams.unitIds;
  const vendorId: string = updateParams.vendorId;
  const vendor = await prisma.vendor.findUnique({ where: { id: vendorId } });

  if (!vendor) {
    return false;
  }

  await prisma.unit.updateMany({
    where: { id: { in: unitIds } },
    data: { vendorId: vendor.id },
  });

  const updated_units = await prisma.unit.findMany({
    where: { id: { in: unitIds }, vendorId: vendor.id },
    select: {
      lp: true,
      vendor: { select: { vendorIdentifier: true } },
    },
  });

  let filtered_updated_units = updated_units.filter(
    (u): u is IUpdatedUnits => u.vendor !== null
  );

  if (filtered_updated_units.length <= 0) return false;

  let res = mappedUnits(filtered_updated_units);
   ```

**Reply by pantharshit00** [reply]
Hello @vvanslyke, did you get a chance to checkout above solution? I think you can also use the new `satisfies` operator here as well now which was recently released.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
